### PR TITLE
Switch memory backed buffers to file backed buffers

### LINKED ISSUE #121: Implement file-backed buffers
Currently data is persisted in memory throughout the entire Receptor pipeline.

This isn't going to work for large payloads that need to be delivered across the mesh, it would be better to persist messages to disk and relay them in chunks over the network if they are on the larger size. Smaller messages could still be delivered as data to the plugin. This could be determined by some metadata about the plugin itself.

Right now we *do* persist data to disk but it's serialized as an individual json payload, this requires reading the entire payload into memory to (de)serialize it.

This also changes the basic message format of what's sent across the wire.

### PR DESCRIPTION
**Pull request by matburt**
This represents work toward #121 

This PR replaces #84

### CODE REVIEW

**Comment by elyezer** [comment]
I was trying to run these changes locally and I am facing the following. The example is with ping command but it also happens when running in either controller or node modes.

```console
$ poetry run receptor \
    -d /tmp/ping-a ping \
    --peer=receptor://127.0.0.1:9999 \
    --delay 0 --count 10 node-a
/home/elyezer/code/receptor/receptor/receptor/connection/sock.py:30: RuntimeWarning: coroutine 'BridgeQueue.__aiter__' was never awaited
  async for chunk in q:
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
ERROR 2020-03-06 14:42:23,757  sock sock.connect
Traceback (most recent call last):
  File "/home/elyezer/code/receptor/receptor/receptor/connection/sock.py", line 44, in connect
    await worker.client(t)
  File "/home/elyezer/code/receptor/receptor/receptor/connection/base.py", line 195, in client
    await self.hello()
  File "/home/elyezer/code/receptor/receptor/receptor/connection/base.py", line 173, in hello
    await self.conn.send(BridgeQueue.one(msg))
  File "/home/elyezer/code/receptor/receptor/receptor/connection/sock.py", line 30, in send
    async for chunk in q:
TypeError: 'async for' received an object from __aiter__ that does not implement __anext__: coroutine
Connection failed. Exiting.
```

**Comment by jhjaggars** [comment]
> I was trying to run these changes locally and I am facing the following. The example is with ping command but it also happens when running in either controller or node modes.
> 
> ```
> $ poetry run receptor \
>     -d /tmp/ping-a ping \
>     --peer=receptor://127.0.0.1:9999 \
>     --delay 0 --count 10 node-a
> /home/elyezer/code/receptor/receptor/receptor/connection/sock.py:30: RuntimeWarning: coroutine 'BridgeQueue.__aiter__' was never awaited
>   async for chunk in q:
> RuntimeWarning: Enable tracemalloc to get the object allocation traceback
> ERROR 2020-03-06 14:42:23,757  sock sock.connect
> Traceback (most recent call last):
>   File "/home/elyezer/code/receptor/receptor/receptor/connection/sock.py", line 44, in connect
>     await worker.client(t)
>   File "/home/elyezer/code/receptor/receptor/receptor/connection/base.py", line 195, in client
>     await self.hello()
>   File "/home/elyezer/code/receptor/receptor/receptor/connection/base.py", line 173, in hello
>     await self.conn.send(BridgeQueue.one(msg))
>   File "/home/elyezer/code/receptor/receptor/receptor/connection/sock.py", line 30, in send
>     async for chunk in q:
> TypeError: 'async for' received an object from __aiter__ that does not implement __anext__: coroutine
> Connection failed. Exiting.
> ```

That's because the protocol for `__aiter__` changed between 3.6 and 3.7 and I goofed it up.  This should be fixed in the latest commit.

**Comment by ghjm** [comment]
I'm getting the following exception on starting up a node with this branch:
```
Task exception was never retrieved
future: <Task finished coro=<Receptor.watch_expire() done, defined at /home/graham/git/receptor/receptor/receptor.py:78> exception=TypeError('an integer is required (got type str)')>
Traceback (most recent call last):
  File "/home/graham/git/receptor/receptor/receptor.py", line 82, in watch_expire
    buffer = self.buffer_mgr[connection["id"]]
  File "/home/graham/git/receptor/receptor/buffers/file.py", line 139, in __missing__
    self[key] = DurableBuffer(self.path, key, self.loop)
  File "/home/graham/git/receptor/receptor/buffers/file.py", line 30, in __init__
    for item in self._read_manifest():
  File "/home/graham/git/receptor/receptor/buffers/file.py", line 64, in _read_manifest
    return json.load(fp)
  File "/usr/lib64/python3.7/json/__init__.py", line 296, in load
    parse_constant=parse_constant, object_pairs_hook=object_pairs_hook, **kw)
  File "/usr/lib64/python3.7/json/__init__.py", line 361, in loads
    return cls(**kw).decode(s)
  File "/usr/lib64/python3.7/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib64/python3.7/json/decoder.py", line 353, in raw_decode
    obj, end = self.scan_once(s, idx)
  File "/home/graham/git/receptor/receptor/serde.py", line 19, in decode_date
    return datetime.datetime.fromtimestamp(o["value"])
TypeError: an integer is required (got type str)
```

**Comment by ghjm** [comment]
Looks like this changes the required worker format, so receptor-http and receptor-sleep (as I have them) are not working:

```
$ receptor send --directive receptor_http:execute foo '{"url": "http://localhost:9000", "method": "GET"}'
'bytes' object has no attribute 'raw_payload'
ERROR 2020-03-10 17:18:53,306  entrypoints EOF was an error result
ERROR: 'bytes' object has no attribute 'raw_payload'
receptor send --directive receptor_sleep:execute foo ''
'NoneType' object has no attribute 'readall'
ERROR 2020-03-10 17:19:33,029  entrypoints EOF was an error result
ERROR: 'NoneType' object has no attribute 'readall'
$ receptor send --directive receptor_sleep:execute foo '{}'
'bytes' object has no attribute 'raw_payload'
ERROR 2020-03-10 17:19:42,695  entrypoints EOF was an error result
ERROR: 'bytes' object has no attribute 'raw_payload'
```

So the only thing I can do with this is ping, but ping isn't working, apparently because I broke it.  See https://github.com/project-receptor/receptor/pull/168.

File: receptor/connection/base.py [hunk 1]
```diff
@@ -28,17 +30,76 @@ def closed(self):
         pass
 
     @abstractmethod
-    async def send(self, bytes_):
+    def send(self, q):
         pass
 
 
-async def watch_queue(conn, buf):
+class BridgeQueue(queue.Queue):
```
Comment (ghjm) [inline receptor/connection/base.py:37]: Should we put this in a separate file and use it from work.py as well?

File: pyproject.toml [hunk 2]
```diff
@@ -8,13 +8,12 @@ description = "A flexible multi-service relayer with remote execution and orches
 license = "Apache-2.0"
 name = "receptor"
 repository = "https://github.com/project-receptor/receptor"
-version = "0.1.0"
+version = "1.0.0"
```
Comment (ghjm) [inline pyproject.toml:11]: I kinda feel like a major version number change should go in its own PR, not be part of a feature PR like this?

File: poetry.lock [hunk 3]
```diff
@@ -165,7 +165,7 @@ description = "Internationalized Domain Names in Applications (IDNA)"
 name = "idna"
 optional = false
 python-versions = ">=2.7, !=3.0.*, !=3.1.*, !=3.2.*, !=3.3.*"
-version = "2.8"
+version = "2.9"
```
Comment (ghjm) [inline poetry.lock:168]: Is this version change (and others in this file) based on some knowledge that we actually need or want this new version?  Do we need to rebuild the python3-idna RPMs?

File: receptor/buffers/file.py [hunk 4]
```diff
@@ -1,84 +1,70 @@
...
 from concurrent.futures import ThreadPoolExecutor
+from json.decoder import JSONDecodeError
 
-import dateutil.parser
-
-from .base import BaseBufferManager
+from .. import serde as json
 
 logger = logging.getLogger(__name__)
 pool = ThreadPoolExecutor()
```
Comment (ghjm) [inline receptor/buffers/file.py:13]: Should this have a configurable number of threads, like the ThreadPoolExecutor in work.py?  Should we perhaps combine forces and have a single ThreadPoolExecutor across our event loop, so we aren't trying to manage CPU execution capacity in two places?

**Comment by matburt** [comment]
> I'm getting the following exception on starting up a node with this branch

@ghjm you'll need to clear your data directory, it's picking up old manifest information before the timestamp change.

File: receptor/connection/base.py (hunk 1 above)
Comment (matburt) [reply receptor/connection/base.py:37]: I think we'll restructure these in the near future but I'd like to get this PR before we do any more refactoring.

File: pyproject.toml (hunk 2 above)
Comment (matburt) [reply pyproject.toml:11]: meh? This is the major thing that lifts this up to 1.0 though.

File: receptor/buffers/file.py (hunk 4 above)
Comment (matburt) [reply receptor/buffers/file.py:13]: Honestly, it's probably good that these have different queues because otherwise one could get starved out, none of them are particularly going to be cpu bound per-se.

Comment (ghjm) [reply receptor/buffers/file.py:13]: As it stands, this does mean the system cannot be processing more than 8 messages in parallel, right?  Perhaps this should just respect the worker threads setting and set itself to the same size.

Comment (matburt) [reply receptor/buffers/file.py:13]: it can't be writing more data payloads to disk than that.

Comment (ghjm) [reply receptor/buffers/file.py:13]: I don't understand why not?

- 10 different peers each send you a 1GB job, all at once
- You grab the 8 threads in the thread pool
- Two of the messages sit and wait

Why can't this happen?

Comment (ghjm) [reply receptor/buffers/file.py:13]: Or are you saying "this _can_ happen, but it's fine because 8 threads is enough to saturate any plausible storage subsystem we might be writing to" ?

**Comment by ghjm** [comment]
I am seeing performance problems with ping.  On the current `devel` branch I get:
```
$ time receptor ping foo --delay 0 --count 1000 > /dev/null

real    0m3.106s
user    0m0.941s
sys     0m2.704s
```

On this branch, I get:
```
$ time receptor ping foo --delay 0 --count 1000 > /dev/null

real    2m8.724s
user    0m0.840s
sys     0m1.838s
```

File: receptor/connection/base.py [hunk 5]
```diff
@@ -28,17 +30,76 @@ def closed(self):
...
+                sleep_time = min(1.0, sleep_time + 0.1)
+
+    @classmethod
+    def one(cls, item):
+        """
+        Constructs a BridgeQueue with the single provided item followed by
+        the sentinel value.  This function does not block.
+        """
+        q = cls()
+        q.put_nowait(item)
```
Comment (ghjm) [inline receptor/connection/base.py:76]: We use put_nowait a lot (this line is just one of many).  What do we expect to happen when the queue is full and cannot accept a new item right now?

File: poetry.lock (hunk 3 above)
Comment (jhjaggars) [reply poetry.lock:168]: No, this is poetry following the rules we set in place for it (i.e. not pinning versions).  This should be set to the versions we have been using.

File: receptor/buffers/file.py (hunk 4 above)
Comment (jhjaggars) [reply receptor/buffers/file.py:13]: I don't know _what_ he was trying to say :), but I agree with the last thing you said.  (Maybe 8 isn't enough to saturate disks we'll see in the wild, but some smallish number of threads probably can)

**Comment by jhjaggars** [comment]
> I am seeing performance problems with ping. On the current `devel` branch I get:
> 
> ```
> $ time receptor ping foo --delay 0 --count 1000 > /dev/null
> 
> real    0m3.106s
> user    0m0.941s
> sys     0m2.704s
> ```
> 
> On this branch, I get:
> 
> ```
> $ time receptor ping foo --delay 0 --count 1000 > /dev/null
> 
> real    2m8.724s
> user    0m0.840s
> sys     0m1.838s
> ```

This is fixed in https://github.com/project-receptor/receptor/pull/130/commits/aa1188bf454f3c494b4695e9f3fd9795196464e0
